### PR TITLE
[RUNTIME] TCP Integration

### PR DESCRIPTION
**Pull request by brandonxxlee**
Adds TCP support for console logging and notifications with Dawn.

TCP process spawns once first message is received from Dawn and will send notifications and console logs via the notification proto over the tcp socekt.

Adds throttling to UDP receive process as well to hopefully improve performance.

Included compiled notification proto for now.

### CODE REVIEW

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: No need to add a newline here

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: You have a call to `select` right under this, which should be blocking. In that case, there should be absolutely no reason to have sleeps anywhere in this loop.

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py:282]: Note that the sleeps in this thread appear to be for throttling purposes, rather than for handling use of non-blocking calls. There should be a comment indicating this.

**Review by nikitakit — CHANGES_REQUESTED** [review]
Just a few minor style comments.

Otherwise code LGTM, but it might be good to get another reviewer because I'm not very familiar with the networking part of the code.

File: runtime/testy/Ansible.py
Comment (yizow) [inline runtime/testy/Ansible.py]: So this line is wrong i think

But re: nikita's comments we shouldn't need this line at all

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py:229]: My previous comment regarding `select` remains unresolved.

**Review by nikitakit — CHANGES_REQUESTED** [review]

File: runtime/testy/Ansible.py
Comment (yizow) [inline runtime/testy/Ansible.py]: We should be using the global variables at the top of the file.

**Review by yizow — CHANGES_REQUESTED** [review]

File: runtime/testy/Ansible.py
Comment (bodiddles) [reply runtime/testy/Ansible.py:229]: Could we use the sleeps for throttling reasons as we will be probably be receiving files faster than we want to be running this thread?

Comment (nikitakit) [reply runtime/testy/Ansible.py:229]: This is a receiver, not a sender. Throttling belongs in sender (in this case, Dawn), not the receiver.

If you want to handle receiving packets faster than this thread can handle, the correct was to implement this is by dropping packets.

Comment (nikitakit) [reply runtime/testy/Ansible.py:229]: Also, we went over this together during the fall semester. The way to implement rate-limited UDP receiver class is:

* Block until there is any data (using `select`)
* Read all available data from the kernel, dropping all but the last packet
* Process the single packet

Do you remember discussing this?

---

The reason sleeping is such a bad idea here is that it forces the kernel packet buffer to fill up, adding system latency equal to the kernel buffer size. If that happens you're going to end up with ridiculously large delays between moving a gamepad and the robot motors reacting. The whole reason we have a stateless UDP system is that any individual packet can be dropped without impacting the system as a whole.

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: EDIT: I misread this the first time.

Please check that `recvfrom` is actually a blocking call.

This condition of `if not unpackagedData` threw me off, and part of it is that unpackage currently doesn't ever return None. I think this condition should be removed, or an exception handler should be added to `unpackage`

File: runtime/testy/Ansible.py
Comment (nikitakit) [reply runtime/testy/Ansible.py:282]: Upon further consideration, I don't think sleeps are the correct throttling mechanism here, either. But I'd be fine with merging this as-is, too, because it's not *too* bad.

Remember how you were debugging an issue where half of the message types were silently being dropped by the kernel? Perhaps that's actually the behavior we want -- your test code might have been spamming things at max speed, but the student code certainly won't be.

File: runtime/testy/Ansible.py
Comment (bodiddles) [reply runtime/testy/Ansible.py]: I changed the code to better reflect this.

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: Delete this, you don't need `UDP_RECV_HZ` since you're not sleeping

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: no need to make this whitespace change

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: Maybe name this `UDP_SEND_PORT`/`UDP_RECV_PORT` to distinguish from TCP now that TCP exists?

**Review by nikitakit — CHANGES_REQUESTED** [review]
Just minor style comments this time.

Maybe @yizow should also review though, since I've had a bad track record of catching networking bugs.

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: You should keep the newline at eof

File: runtime/testy/Ansible.py
Comment (nikitakit) [inline runtime/testy/Ansible.py]: gratuitous whitespace change

**Review by nikitakit — APPROVED** [review]
lgtm now, ideally @yizow would also give this a look

File: runtime/testy/Ansible.py
Comment (yizow) [reply runtime/testy/Ansible.py:282]: The problem with throttling the sending liek this is that if whatever is on the other end of the `pipe` is generating data faster than this thread processes, its gonna fill up and cause a constant latency, as everything on the other end starts blocking on this function. 

Or, if we use non-blocking, the stuff on the other end will error out with `pipe full` errors. 
Especially since this is TCP, it doesn't seem like we want to drop any of these packets. 

It looks like this is purely used for 
1) confirming student code uploaded
2) sending over console output

The former is low frequency/ bandwidth and likely won't cause an issue.
The latter is potentially high frequency, high bandwidth, and most likely WILL cause an issue. Especially if you're throttling.

Maybe batching messages in the pipe together and sending them that way? 


In the interest of getting this merged though, I'm also okay with merging this.
But the console logging commit better fix it.

**Review by yizow — APPROVED** [review]

**Pull request merged**
